server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,345 @@
+            } else {
+                // If file has been uploaded to the Remote Store, fetch it from the Remote Store in blocks via
+                // OnDemandCompositeBlockIndexInput
+                logger.trace("File to be fetched from Remote");
+                return new OnDemandCompositeBlockIndexInput(remoteDirectory, name, localDirectory, fileCache, context);
+            }
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Acquires and returns a {@link Lock} for a file with the given name.
+     * @param name the name of the lock file
+     * @throws LockObtainFailedException (optional specific exception) if the lock could not be
+     * obtained because it is currently held elsewhere.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public Lock obtainLock(String name) throws IOException {
+        logger.trace("obtainLock() called {}", name);
+        writeLock.lock();
+        try {
+            return localDirectory.obtainLock(name);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Closes the directory
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void close() throws IOException {
+        writeLock.lock();
+        try {
+            localDirectory.close();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Returns a set of files currently pending deletion in this directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public Set<String> getPendingDeletions() throws IOException {
+        readLock.lock();
+        try {
+            return localDirectory.getPendingDeletions();
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * Function to perform operations once files have been uploaded to Remote Store
+     * Currently deleting the local files here, as once uploaded to Remote, local files are safe to delete
+     * @param files : recent files which have been successfully uploaded to Remote Store
+     * @throws IOException in case of I/O error
+     */
+    public void afterSyncToRemote(Collection<String> files) throws IOException {
+        logger.trace("afterSyncToRemote called for {}", files);
+        if (remoteDirectory == null) {
+            logger.trace("afterSyncToRemote called even though remote directory is not set");
+            return;
+        }
+        for (String fileName : files) {
+            writeLock.lock();
+            try {
+                localDirectory.deleteFile(fileName);
+            } finally {
+                writeLock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Return the list of files present in Remote
+     */
+    private String[] getRemoteFiles() {
+        String[] remoteFiles;
+        try {
+            remoteFiles = remoteDirectory.listAll();
+        } catch (Exception e) {
```

Comment:
Can we handle specific exceptions instead of a generic catch clause? 

What if a call to `remoteDirectory` fails due to an IOException?

Reply:
+1